botorch/acquisition/max_value_entropy_search.py

```diff
@@ -154,62 +119,178 @@ def forward(self, X: Tensor) -> Tensor:
-        check_no_nans(mean)
-        check_no_nans(variance)
-
         ig = self._compute_information_gain(
             X=X, mean_M=mean, variance_M=variance, covar_mM=variance.unsqueeze(-1)
         )
+        return ig.mean(dim=0)  # average over fantasies
 
-        ig = ig.mean(dim=0)  # average over fantasies (when used)
+    def set_X_pending(self, X_pending: Optional[Tensor] = None) -> None:
+        r"""Set pending design points.
 
-        return ig
+        Set "pending points" to inform the acquisition function of the candidate
+        points that have been generated but are pending evaluation.
+
+        Args:
+            X_pending: `n x d` Tensor with `n` `d`-dim design points that have
+                been submitted for evaluation but have not yet been evaluated.
+        """
+        if X_pending is not None:
+            X_pending = X_pending.detach().clone()
+        self._sample_max_values(num_samples=self.num_mv_samples, X_pending=X_pending)
+        self.X_pending = X_pending
+
+    # ------- Abstract methods that need to be implemented by subclasses ------- #
 
     @abstractmethod
     def _compute_information_gain(
         self, X: Tensor, mean_M: Tensor, variance_M: Tensor, covar_mM: Tensor
     ) -> Tensor:
-        r"""Computes the information gain at the design points `X`.
+        r"""Compute the information gain at the design points `X`.
+
+        `num_fantasies = 1` for non-fantasized models.
 
          Args:
             X: A `batch_shape x 1 x d`-dim Tensor of `batch_shape` t-batches
                 with `1` `d`-dim design point each.
-            mean_M, variance_M: `batch_shape x num_fantasies`-dim Tensors of
-                `batch_shape` t-batches with `num_fantasies` fantasies.
-                `num_fantasies = 1` for non-fantasized models.
-                All are obtained without noise.
-            covar_mM: `batch_shape x num_fantasies x (1 + num_trace_observations)`
-                -dim Tensor. `num_fantasies = 1` for non-fantasized models.
-                All are obtained without noise.
+            mean_M: A `batch_shape x num_fantasies`-dim Tensor of means.
+            variance_M: A `batch_shape x num_fantasies`-dim Tensor of variances.
+            covar_mM: A `batch_shape x num_fantasies x (1 + num_trace_observations)`
+                -dim Tensor of covariances.
 
         Returns:
-            A `batch_shape`-dim Tensor of information gains at the
-            given design points `X`.
+            A `num_fantasies x batch_shape`-dim Tensor of information gains at the
+            given design points `X` (`num_fantasies=1` for non-fantasized models).
         """
         pass  # pragma: no cover
 
-    def set_X_pending(self, X_pending: Optional[Tensor] = None) -> None:
-        r"""Informs the acquisition function about pending design points.
+    @abstractmethod
+    def _sample_max_values(
+        self, num_samples: int, X_pending: Optional[Tensor] = None
```

Comment:
1. We could, but then it may require folks implementing new methods to go search for it. This way we don't necessarily require `self.num_samples` to be set. For instance we could then implement some algos that adaptively change the number of max value samples in some fashion...
2. Yes.